app/gui2/src/util/callTree.ts

```diff
@@ -0,0 +1,152 @@
+import { Tree } from '@/generated/ast'
+import type { SuggestionEntryArgument } from '@/stores/suggestionDatabase/entry'
+import { tryGetIndex } from './array'
+import type { AstExtended } from './ast'
+
+/**
+ * Information about an argument that doesn't have an assigned value yet, therefore are not
+ * represented in the AST.
+ */
+export class ArgumentPlaceholder {
+  constructor(
+    public index: number,
+    public info: SuggestionEntryArgument,
+  ) {}
+
+  static isInstance(obj: unknown): obj is ArgumentPlaceholder {
```

Comment:
consider using the spec-defined property `[Symbol.hasInstance]` instead, and then just doing a regular `instanceof` check

Reply:
The point of this method is to have access to this predicate as a standalone symbol, so I can use it in the widget matchers. I guess I could make a function `isInstance<T>(ctor: { constructor(...args: any[]): T }): (obj: unknown) -> obj instanceof T` or something like that, but the way it is now seems to be a bit simpler.

Reply:
`ArgumentPlaceholder[Symbol.hasInstance]` should work? AFAICT there's nothing stopping you from declaring the type as a predicate: [Playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8BQ1oQC5nQS2NA2gMoCeAtgEYD2IAdABaQCSAdhmM8AKYC6AFAG5gQAV04AuaMOYBrZpQDuzAJQTBIztGwx4yVKgBOndMP3Noa0ZtaYOnSgDM4KaAF8kLoA)

Reply:
I think implementing `Symbol.hasInstance` by using `instanceof` will potentially cause a stack overflow?

Reply:
or `[Symbol.hasInstance] = Object[Symbol.hasInstance] as <predicate type here>` - it's less type-safe but predicates are not really type-safe anyway

Reply:
I've created my own generic `isInstance` function, so it can be curried, which is necessary for widget type predicates.